Modules/mechaclient.py

```diff
@@ -0,0 +1,225 @@
+
+        """
+        self._api_handler = None  # TODO: replace with handler init once it exists
+        self._database_manager = None  # TODO: replace with dbm once it exists
+        self._rat_cache = None  # TODO: replace with ratcache once it exists
+        self.graceful_errors = handle_errors_gracefully
+        # Event.events['on_connect'].decorated_coro = self.on_connect
+        # Event.events['on_message'].decorated_coro = self.on_message
+        super().__init__(*args, **kwargs)
+
+    async def on_connect(self):
+        """
+        Called upon connection to the IRC server
+        :return:
+        """
+        log.debug(f"Connecting to channels...")
+        # join a channel
+        for channel in config["irc"]["channels"]:
+            log.debug(f"Configured channel {channel}")
+            await self.join(channel)
+
+        log.debug("joined channels.")
+
+        # do event callback
+        await events.on_connect.emit(self)
+        # call the super
+        super().on_connect()
+
+    async def on_message(self, channel, user, message: str) -> any:
+        """
+        Triggered when a message is received
+        :param channel: Channel the message arrived in
+        :param user: user that triggered the message
+        :param message: message body
+        :return:
+        """
+        log.debug(f"{channel}: <{user}> {message}")
+
+        try:
+            # build a context object
+            ctx = await Context.from_message(self, channel, user, message)
+
+            # and emit the raw message
+            await events.on_message_raw.emit(context=ctx)
+
+            if ctx.user.nickname == config['irc']['nickname']:
+                # don't do this and the bot can get into an infinite
+                # self-stimulated positive feedback loop.
+                log.debug(f"Ignored {message} (anti-loop)")
+                return None
+
+            if not message.startswith(prefix):
+                # prevent bot from processing commands without the set prefix
+                log.debug(f"Ignored {message} (not a command)")
+                return await events.on_message.emit(context=ctx)
+
+            else:  # await command execution
+                # sanitize input string headed to command executor
+                sanitized_message = sanitize(message)
+                log.debug(f"Sanitized {sanitized_message}, Original: {message}")
+                return await events.on_command.emit(context=ctx)
+
+        except Exception as exc:
+            ex_uuid = uuid4()
+            log.exception(ex_uuid)
+            error_message = graceful_errors.make_graceful(exc, ex_uuid)
+
+            # check to see if we are gracefully handling errors
+            if self.graceful_errors:
+                # and report it to the user
+                await self.message(channel, error_message)
+            else:
+                # we are not handling things gracefully, raise.
+                raise
```

Comment:
![grafik](https://user-images.githubusercontent.com/29020266/47817260-8057da80-dd55-11e8-911a-09cac484a515.png)

Well, `exc`. But still.

Reply:
uhhh.... im not sure what you are telling me here. sowhat does `e` have to do with all this?

Reply:
Sorry, I couldn't help myself. The `exc` is missing from after the `raise`.

Reply:
`raise` will implicitly raise the caught exception
i don't believe i need to explicitly tell python to `raise exc`

Reply:
Urgh, so using `raise` without arguments re-raises the currently handled exception? I didn't know that. Could you do me a favour and add the `exc` anyhow? I might not be the only one bamboozled by this syntax.

Reply:
```py
try:
    0/0
except ZeroDivisionError:
    print("someone divided by zero (Again...)")
    raise
Traceback (most recent call last):
someone divided by zero (Again...)
  File "J:\worspaces\pycharm\fuelrats\pipsqueak3\mecha_py_3_7\lib\site-packages\IPython\core\interactiveshell.py", line 2961, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-479d368a3351>", line 3, in <module>
    0/0
ZeroDivisionError: division by zero
```

Reply:
I believe you, don't worry. Though I believe explicitly providing the exception to raise would prevent misunderstandings, like this one.

Reply:
hmm, i beg to differ. if i included an Exception class / object here it might be mis-read as me raising a *new* exception. or python would see it as a an exception during exception handling which would be worse on the stack trace.

By just having a bare `raise` it should be clear that i am raising the exception that was caught as im not raising anything else. IMHO